src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,193 @@
+
+import (
+	"autoscaler/db"
+	"autoscaler/metricsforwarder/forwarder"
+	"autoscaler/models"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"code.cloudfoundry.org/cfhttp/handlers"
+	"code.cloudfoundry.org/lager"
+	"github.com/patrickmn/go-cache"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type CustomMetricsHandler struct {
+	metricForwarder forwarder.MetricForwarder
+	policyDB        db.PolicyDB
+	credentialCache cache.Cache
+	cacheTTL        time.Duration
+	logger          lager.Logger
+}
+
+func NewCustomMetricsHandler(logger lager.Logger, metricForwarder forwarder.MetricForwarder, policyDB db.PolicyDB, credentialCache cache.Cache, cacheTTL time.Duration) *CustomMetricsHandler {
+	return &CustomMetricsHandler{
+		metricForwarder: metricForwarder,
+		policyDB:        policyDB,
+		credentialCache: credentialCache,
+		cacheTTL:        cacheTTL,
+		logger:          logger,
+	}
+}
+
+func (mh *CustomMetricsHandler) PublishMetrics(w http.ResponseWriter, r *http.Request, vars map[string]string) {
+	w.Header().Set("Content-Type", "application/json")
+
+	username, password, authOK := r.BasicAuth()
+
+	if authOK == false {
+		mh.logger.Info("error-processing-authorizaion-header")
+		handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+			Code:    "Authorization-Failure-Error",
+			Message: "Error varifying user credentials. Basic authorization is not used properly"})
+		return
+	}
+
+	var isValid bool
+	var credentials models.CustomMetricCredentials
+
+	appID := vars["appid"]
+	res, found := mh.credentialCache.Get(appID)
+	if found {
+		// Credentials found in cache
+		credentials = res.(models.CustomMetricCredentials)
+		isValid = mh.validateCredentials(username, credentials.Username, password, credentials.Password)
+	}
+
+	// Credentials not found in cache or
+	// stale cache entry with invalid credential found in cache
+	// search in the database and update the cache
+	if !found || !isValid {
+		usernameHash, passwordHash, err := mh.policyDB.GetCustomMetricsCreds(appID)
+		if err != nil {
+			mh.logger.Error("error-during-getting-binding-credentials-from-policyDB", err)
+			handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+				Code:    "Interal-Server-Error",
+				Message: "Error reading custom metrics request body"})
+			return
+		}
+		credentials = models.CustomMetricCredentials{
+			Username: usernameHash,
+			Password: passwordHash,
+		}
+		isValid = mh.validateCredentials(username, credentials.Username, password, credentials.Password)
+		// If Credentials in DB is not valid
+		if !isValid {
+			mh.logger.Error("error-validating-authorizaion-header", err)
+			handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+				Code:    "Authorization-Failure-Error",
+				Message: "Basic authorization credential does not match"})
+			return
+		}
+		// Valid credentials found in DB so update the cache
+		mh.credentialCache.Set(appID, credentials, mh.cacheTTL)
```

Comment:
I think we need to update credentialCache with "usernameHash, passwordHash"  regardingless whether the validation result of the auth  info in incoming request. 
"usernameHash, passwordHash" is retrieved from database, so these values should be correct,

Reply:
Done